File: solution files/Javascript/part-01/task-04-categorize-secrets/blueprint_decoder.js
const fs = require("fs");

function decodeBlueprint(filename) {
    const content = fs.readFileSync(filename, "utf8");
    const pattern = /\{\* (.*?) \*\}/g;
    const matches = [...content.matchAll(pattern)];
    const secrets = matches.map((match) => match[1]);
    return secrets;
}

// Enhanced version with error handling
// If file doesn't exist, return empty array and print error message
function decodeBlueprintSafe(filename) {
    try {
        return decodeBlueprint(filename);
    } catch (err) {
        console.log(`Error: File '${filename}' not found.`);
        return [];
    }
}

// Function to format and display secrets in a professional report
// Includes header, separator lines, numbered list, and footer
function displaySecretsReport(secrets) {
    const separator = "=".repeat(50);
    console.log("\n" + separator);
    console.log("DECODED SECRETS REPORT".padEnd(50));
    console.log(separator);
    console.log(`Total secrets found: ${secrets.length}\n`);
    secrets.forEach((secret, index) => {
        const num = String(index + 1).padStart(2, " ");
        console.log(`  [${num}] ${secret}`);
    });
    console.log("\n" + separator + "\n");
}

// Function to categorize secrets by their type (word before the colon)
function categorizeSecrets(secrets) {
    const categories = {};
    for (const secret of secrets) {
        const category = secret.includes(":") ? secret.split(":")[0].trim() : "UNCLASSIFIED";
        categories[category] = (categories[category] ?? 0) + 1;
    }
    return categories;
}

// Use it
const secrets = decodeBlueprintSafe("blueprint-data.txt");
displaySecretsReport(secrets);

const categories = categorizeSecrets(secrets);
console.log("\nSecret Categories:");
Object.keys(categories)
    .sort()
    .forEach((category) => {
        console.log(`  ${category}: ${categories[category]}`);
    });

module.exports = {
    decodeBlueprint,
    decodeBlueprintSafe,
    displaySecretsReport,
    categorizeSecrets,
};

File: starter-files/Python/part-03/task-01-refactor/blueprint_decoder.py
def decode_blueprint(filename):
    try:
        with open(filename, "r") as f:
            content = f.read()
        import re
        secrets = re.findall(r"\{\* (.*?) \*\}", content)
        print("=" * 40)
        print("DECODED SECRETS REPORT")
        print("=" * 40)
        print(f"Found {len(secrets)} secret(s):\n")
        for i, s in enumerate(secrets, 1):
            print(f"{i}. {s}")
        print("=" * 40)
        cats = {}
        for s in secrets:
            key = s.split(":")[0].strip() if ":" in s else "UNKNOWN"
            cats[key] = cats.get(key, 0) + 1
        return cats
    except FileNotFoundError:
        print(f"Error: '{filename}' not found.")
        return {}
